### ISSUE TITLE
UpdateFromTestsRepo uses a single mutex for all updates

### ISSUE THREAD

**Issue by meling** [quickfeed/quickfeed#818]
This may delay processing for courses that update their `tests` repo concurrently. We should consider adding a mutex per course, to allow concurrency per course.

**Comment by JoachimTislov** [comment]
> This may delay processing for courses that update their `tests` repo concurrently. We should consider adding a mutex per course, to allow concurrency per course.

Yes this can delay processing, the `updateMutex` variable is to my understanding global and stored on the heap, which will only allow one update to happen at time.

The function `UpdateFromTestsRepo` is executed in these two cases:
- A teacher pushes an update to the tests repository, the Github Webhook then triggers a creation of a `go routine`, and updates the semaphore limit, which is initialized as `const maxConcurrentTestRuns = 5`
    
  ```go
  go func() {
  	wh.sem <- struct{}{} // acquire semaphore
  	wh.handlePush(e)
  	<-wh.sem // release semaphore
  	// remove commitID from duplicate map (to avoid memory leak)
  	wh.dup.Remove(commitID)
  }()
  ```
- A teacher can also request an update in Quickfeed's web interface;
![Image](https://github.com/user-attachments/assets/07383147-94c9-49e4-b9d2-111913d4d78c)
This will be call the [`UpdateAssignments`](https://github.com/quickfeed/quickfeed/blob/master/web/quickfeed_service.go#496) API, which will run `UpdateFromTestsRepo` in the main thread.

If teachers sends multiple requests to Quickfeed by either the API or through the Github webhook, will it result in a queue and the response for teacher using the API will be delayed. The `push` for tests repositories will eventually be handled.

That said, theres a limited amount of courses and the case of multiple teachers wanting to update their tests is rare, but configuring a `mutex` per course is reasonable, and will improve response time.
